onprc_ehr/src/org/labkey/onprc_ehr/issues/RestrictedIssueProviderImpl.java

```diff
@@ -0,0 +1,136 @@
+    {
+        setPropertyValue(c, issueDefName, PROP_RESTRICTED_ISSUE_LIST_GROUP, null != group ? String.valueOf(group.getUserId()) : "0");
+    }
+
+    @Override
+    public @Nullable Group getRestrictedIssueListGroup(Container c, String issueDefName)
+    {
+        String groupId = getPropertyValue(c, issueDefName, PROP_RESTRICTED_ISSUE_LIST_GROUP);
+        if (null == groupId)
+            return null;
+
+        return SecurityManager.getGroup(Integer.valueOf(groupId));
+    }
+
+    private void setPropertyValue(Container c, String issueDefName, String key, String value)
+    {
+        PropertyManager.PropertyMap props = PropertyManager.getWritableProperties(c, getPropMapName(issueDefName), true);
+        props.put(key, value);
+        props.save();
+    }
+
+    private String getPropertyValue(Container c, String issueDefName, String key)
+    {
+        return PropertyManager.getProperties(c, getPropMapName(issueDefName)).get(key);
+    }
+
+    private String getPropMapName(String issueDefName)
+    {
+        if (issueDefName == null)
+            throw new IllegalArgumentException("Issue def name must be specified");
+
+        return CAT_RESTRICTED_ISSUE_DEF_PROPERTIES + issueDefName;
+    }
+
+    @Override
+    public boolean hasPermission(User user, @NotNull Issue issue, List<Issue> relatedIssues, List<ValidationError> errors)
+    {
+        Container issueContainer = ContainerManager.getForId(issue.getContainerId());
+        if (issueContainer != null && isRestrictedIssueTracker(issueContainer, issue.getIssueDefName()))
+        {
+            Group group = getRestrictedIssueListGroup(issueContainer, issue.getIssueDefName());
+            if (!checkAccess(user, issue, group))
+            {
+                errors.add(new SimpleValidationError("This issue is in a restricted issue list. You do not have access to this issue"));
+                return false;
+            }
+        }
+
+        // the user must also have access to all related issues
+        for (Issue related : relatedIssues)
+        {
+            Container relatedContainer = ContainerManager.getForId(related.getContainerId());
+            if (relatedContainer != null && isRestrictedIssueTracker(relatedContainer, related.getIssueDefName()))
+            {
+                Group relatedGroup = getRestrictedIssueListGroup(relatedContainer, related.getIssueDefName());
+                if (!checkAccess(user, related, relatedGroup))
+                {
+                    errors.add(new SimpleValidationError(String.format("A related issue : %d is in a restricted issue list. You do not have access to that issue", related.getIssueId())));
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private boolean checkAccess(User user, @NotNull Issue issue, @Nullable Group groupWithAccess)
+    {
+        // Site admins always have access
+        if (user.isInSiteAdminGroup())
+            return true;
```

Comment:
could move this up to the start of the `hasPermission` method and short circuit all of the issue specific checks if the user is a site admin.

Reply:
sure, that makes sense